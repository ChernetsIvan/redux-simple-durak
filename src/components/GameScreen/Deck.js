import React from "react";
import PropTypes from "prop-types";

import Card from "./Card/Card";
import HiddenCard from "./Card/HiddenCard";

const Deck = ({cards}) => {
    var cardsWithoutLast = [];
    cardsWithoutLast = cards.concat();
    let trumpCard = cardsWithoutLast.pop();

    //1)Козырная карта.
    let output_TrumpCard = null;
    if (trumpCard !== undefined) {
      //"Переопределяем" bootstrap-класс col-6 вместо col-1:
      output_TrumpCard = (
        <Card
          rank={trumpCard.rank}
          suit={trumpCard.suit}
          bootStrapColClass="col-6"
          hidden={false}
        />
      );
    }

    //2)Имитация колоды.
    let output_MockDeck = null;
    let output_CountOfCardsInDeck = null;
    if (cardsWithoutLast.length > 0) {
      output_MockDeck = (
        <HiddenCard bootStrapColClass="col-6" />
      );
      output_CountOfCardsInDeck = cardsWithoutLast.length;
    }

    let output = (
      <div className="container">
        <div className="row">
          {output_TrumpCard}
          {output_MockDeck}
        </div>
        <div className="row">
          <div className="col-6" />
          <div className="col-6">
            <div className="container">
              <div className="row justify-content-center">
                <div className="col-auto">
                  <b>{output_CountOfCardsInDeck}</b>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    );
    return output;
};

export default Deck;

Deck.propTypes = {
  cards: PropTypes.array
};
