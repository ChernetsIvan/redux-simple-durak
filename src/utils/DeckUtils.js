import { pull, find } from 'lodash-es';

import {cardRanks, cardSuits} from './../constants/GameConstants';

import CardModel from './../models/CardModel';

class DeckUtils{
    //отсортировать входной массив карт по возрастанию isAsc=true
    //(убыванию-> isAsc=false)  по параметру 'power'
    static sortInputDeckByPower(inputDeck, isAsc){
        inputDeck.sort(function(a,b){
            if(isAsc){
                return a.power - b.power;
            }else{
                return b.power - a.power;
            }        
        });
    }

    //Возвращает наполненную всеми 36-ю картами колоду
    static fillCards(trumpSuit){
        let fullDeck = [];
        //Устанавливает power для каждой карты по принципу:
        //козырь->10*rank.cardValue, НЕкозырь->1*rank.cardValue
        let z = 0;
        let power, coef;               
        cardSuits.forEach(function(element,index, array){
            if(element.suit === trumpSuit.suit){
                coef = 10;
            }else{
                coef = 1;
            }
            cardRanks.forEach(function(element2, index2, array2){
                power = coef * element2.cardValue;
                fullDeck.push(
                    new CardModel(z, element2, element, power)
                );
                z++;
            });
        });
        return fullDeck;
    }

    //Возвращает перемешанную колоду
    static shuffle(inputDeck){
        var newFullDeck = [];
        while(inputDeck.length > 0){
            //Случайный индекс в inputDeck
            let randIndex = Math.random() * (inputDeck.length-1);
            randIndex = Math.round(randIndex);

            let card = inputDeck[randIndex];
            newFullDeck.push(card);

            inputDeck.splice(randIndex, 1);
        }
        return newFullDeck;
    }

    //Перемещает первую попавшуюся карту с козырной мастью
    //в конец колоды 
    static moveAnyCardWithTrumpSuitToTailOfFullDeck(trumpSuit, fullDeck){
        let card = find(fullDeck, (card) => {
            return card.suit.suit === trumpSuit.suit;
        });
        pull(fullDeck, card);
        fullDeck.push(card);
    }

    //взять из НАЧАЛА (!) колоды fullDeck и положить в inputDeck столько карт, 
    //чтобы в inputDeck в итоге было 6 карт
    static giveUpToSixCards(fullDeck, inputDeck){
        //если хотя бы одно из условий не выполняется, в цикл не заходим:
        while(inputDeck.length < 6 && fullDeck.length > 0){
            let card = fullDeck.shift();
            inputDeck.push(card);
        }
    }

    //Возвращает козырную масть (не карту!) для будущей игры
    static chooseSuitForTrumpCardInFuture(){
        let randIndex = Math.random() * (cardSuits.length-1);
        randIndex = Math.round(randIndex);
        return cardSuits[randIndex];
    }
}

export default DeckUtils;