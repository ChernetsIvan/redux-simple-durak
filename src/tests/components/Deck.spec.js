import React from "react";
import renderer from "react-test-renderer";
import Deck from "./../../components/GameScreen/Deck";
import DeckUtils from "./../../utils/DeckUtils";

test("Deck snapshot testing", () => {
  const cards = DeckUtils.fillCards({ suit: "Ð‘" });
  const component = renderer.create(<Deck cards={cards.slice(0,10)} />);
  expect(component.toJSON()).toMatchSnapshot();
});
