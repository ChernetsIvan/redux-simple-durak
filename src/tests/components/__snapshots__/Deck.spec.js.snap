// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deck testing snapshot testing with cards 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col-6 card mb-1"
      onClick={[Function]}
    >
      <div
        className="row"
      >
        <b
          className="ml-1"
        >
          <b
            className="blackText"
          >
            6
          </b>
        </b>
      </div>
      <div
        className="row"
      >
        <img
          alt=""
          className="littleSuitImg ml-1"
          src="club.png"
        />
      </div>
      <div
        className="row justify-content-center"
      >
        <img
          alt=""
          className="bigSuitImg my-2"
          src="club.png"
        />
      </div>
      <div
        className="row justify-content-end"
      >
        <img
          alt=""
          className="littleSuitImg rotated mr-1"
          src="club.png"
        />
      </div>
      <div
        className="row justify-content-end"
      >
        <b
          className="rotated mr-1"
        >
          <b
            className="blackText"
          >
            6
          </b>
        </b>
      </div>
    </div>
    <div
      className="col-6 card mb-1 shirt"
    />
  </div>
  <div
    className="row"
  >
    <div
      className="col-6"
    />
    <div
      className="col-6"
    >
      <div
        className="container"
      >
        <div
          className="row justify-content-center"
        >
          <div
            className="col-auto"
          >
            <b>
              9
            </b>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Deck testing snapshot testing without cards 1`] = `
<div
  className="container"
>
  <div
    className="row"
  />
  <div
    className="row"
  >
    <div
      className="col-6"
    />
    <div
      className="col-6"
    >
      <div
        className="container"
      >
        <div
          className="row justify-content-center"
        >
          <div
            className="col-auto"
          >
            <b />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
